### ISSUE TITLE
The symbol information ignores additional restrictions on range provided by surrounding DW_TAG_lexical_block

### ISSUE THREAD

**Issue by wcohen** [dyninst/dyninst#1225]
**Intention**
I have some tooling note when multiple variables share the same register due to optimization (https://github.com/wcohen/quality_info/blob/master/dyninsttools/cse.C).  It outputs the ranges where there the variables are live and register represents more than one variable. 

**Describe the bug**
I noticed for /usr/lib/debug/lib/modules/5.16.12-200.fc35.x86_64/vmlinux there were some overlaps reported that had ranges that started at 0 or ended at 0xfffffffffffffff.  Like the following for fault_in_readablet:

```
fault_in_readable
[0,ffffffffffffffff) x86_64::rdx __val_gu __val_gu 
[ffffffff812c9e30,ffffffff812c9e50) x86_64::rdi uaddr start 
[ffffffff812c9ea5,ffffffff812c9ead) x86_64::rdi uaddr start 
[ffffffff812c9ec5,ffffffff812c9ec8) x86_64::rdi uaddr start 
```
Looked at the resulting  debuginfo from the vmlinux using:
`llvm-dwarfdump -c --name fault_in_readable /usr/lib/debug/lib/modules/5.16.12-200.fc35.x86_64/vmlinux > fault_in_readable
`

It becomes clear what is happening.  A number of  DW_TAG_lexical_block tags limit the scope of the variables and dyninst is not using the DW_TAG_lexical_block to limit the range of those variables.   One __val_gu defined by:

```
0x02bc0bd0:   DW_TAG_lexical_block
                DW_AT_low_pc	(0xffffffff812c9ea5)
                DW_AT_high_pc	(0xffffffff812c9eb1)
                DW_AT_sibling	(0x02bc0c09)
```
```
0x02bc0bf9:     DW_TAG_variable
                  DW_AT_name	("__val_gu")
                  DW_AT_decl_file	("/usr/src/debug/kernel-5.16.12/linux-5.16.12-200.fc35.x86_64/mm/gup.c")
                  DW_AT_decl_line	(1800)
                  DW_AT_decl_column	(0x07)
                  DW_AT_type	(0x02ba4d2f "unsigned char")
                  DW_AT_location	(DW_OP_reg1 RDX)

```
And the other is defined by
```
0x02bc0c09:   DW_TAG_lexical_block
                DW_AT_low_pc	(0xffffffff812c9e7b)
                DW_AT_high_pc	(0xffffffff812c9e87)

```
```
0x02bc0c2e:     DW_TAG_variable
                  DW_AT_name	("__val_gu")
                  DW_AT_decl_file	("/usr/src/debug/kernel-5.16.12/linux-5.16.12-200.fc35.x86_64/mm/gup.c")
                  DW_AT_decl_line	(1808)
                  DW_AT_decl_column	(0x07)
                  DW_AT_type	(0x02ba4d2f "unsigned char")
                  DW_AT_location	(DW_OP_reg1 RDX)
```

In this particular case the __val_gu is coming from two different instances of the [do_get_user_call](https://elixir.bootlin.com/linux/v5.16.12/C/ident/do_get_user_call) macro.

They don't actually overlap based on the DW_TAG_lexical_block information.  However, dyninst isn't incorporating the DW_TAG_lexical_block information into the information about the variable ranges.

**To Reproduce**
Will work on a smaller reproducer.

**Expected behavior**
For the paricular fault_in_readable function no overlap reported for __val_gu variables. by cse.

**System (please complete the following information):**
 - x86_64 Fedora 35
 - dyninst-11.0.1-3.fc35.x86_64
 - glibc-2.34-28.fc35.x86_64

**Additional context**
The issue still exists in the newly release dyninst-12.1.0.

**Comment by wcohen** [comment]
Have a small reproducer for the issue, pr1225.c:

```
#include <stdio.h>
int main(int argc, char *argv[])
{
	{register int a; a=argc; printf("%d\n", a);}
	printf("separate statements\n");
	{register int a; a=argc; printf("%d\n", a);}
	return 0;
}
```
$ gcc -g pr1225.c -o pr1225
$ ~/research/profiling/debuginfo/quality_info/dyninsttools/cse pr1225
unable to find puts
unable to find printf
main
[0,ffffffffffffffff) x86_64::rbx a a 

The dwarf information definitely has these as non-overlapping when one looks at the DW_TAG_lexical_block for each variable a:
```
$ llvm-dwarfdump -c -name main pr1225
pr1225:	file format elf64-x86-64

0x00000094: DW_TAG_subprogram
              DW_AT_external	(true)
              DW_AT_name	("main")
              DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
              DW_AT_decl_line	(2)
              DW_AT_decl_column	(0x05)
              DW_AT_prototyped	(true)
              DW_AT_type	(0x00000058 "int")
              DW_AT_low_pc	(0x0000000000401136)
              DW_AT_high_pc	(0x0000000000401183)
              DW_AT_frame_base	(DW_OP_call_frame_cfa)
              DW_AT_call_all_tail_calls	(true)
              DW_AT_sibling	(0x0000010d)

0x000000b6:   DW_TAG_formal_parameter
                DW_AT_name	("argc")
                DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                DW_AT_decl_line	(2)
                DW_AT_decl_column	(0x0e)
                DW_AT_type	(0x00000058 "int")
                DW_AT_location	(DW_OP_fbreg -36)

0x000000c3:   DW_TAG_formal_parameter
                DW_AT_name	("argv")
                DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                DW_AT_decl_line	(2)
                DW_AT_decl_column	(0x1a)
                DW_AT_type	(0x0000010d "char**")
                DW_AT_location	(DW_OP_fbreg -48)

0x000000d0:   DW_TAG_lexical_block
                DW_AT_low_pc	(0x0000000000401146)
                DW_AT_high_pc	(0x000000000040115a)
                DW_AT_sibling	(0x000000f0)

0x000000e5:     DW_TAG_variable
                  DW_AT_name	("a")
                  DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                  DW_AT_decl_line	(4)
                  DW_AT_decl_column	(16)
                  DW_AT_type	(0x00000058 "int")
                  DW_AT_location	(DW_OP_reg3 RBX)

0x000000ef:     NULL

0x000000f0:   DW_TAG_lexical_block
                DW_AT_low_pc	(0x0000000000401164)
                DW_AT_high_pc	(0x0000000000401178)

0x00000101:     DW_TAG_variable
                  DW_AT_name	("a")
                  DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                  DW_AT_decl_line	(6)
                  DW_AT_decl_column	(16)
                  DW_AT_type	(0x00000058 "int")
                  DW_AT_location	(DW_OP_reg3 RBX)

0x0000010b:     NULL

0x0000010c:   NULL
```

**Comment by wcohen** [comment]
Have a small reproducer for the issue, pr1225.c:

```
#include <stdio.h>
int main(int argc, char *argv[])
{
	{register int a; a=argc; printf("%d\n", a);}
	printf("separate statements\n");
	{register int a; a=argc; printf("%d\n", a);}
	return 0;
}
```
$ gcc -g pr1225.c -o pr1225
$ ~/research/profiling/debuginfo/quality_info/dyninsttools/cse pr1225
unable to find puts
unable to find printf
main
[0,ffffffffffffffff) x86_64::rbx a a 

The dwarf information definitely has these as non-overlapping when one looks at the DW_TAG_lexical_block for each variable a:
```
$ llvm-dwarfdump -c -name main pr1225
pr1225:	file format elf64-x86-64

0x00000094: DW_TAG_subprogram
              DW_AT_external	(true)
              DW_AT_name	("main")
              DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
              DW_AT_decl_line	(2)
              DW_AT_decl_column	(0x05)
              DW_AT_prototyped	(true)
              DW_AT_type	(0x00000058 "int")
              DW_AT_low_pc	(0x0000000000401136)
              DW_AT_high_pc	(0x0000000000401183)
              DW_AT_frame_base	(DW_OP_call_frame_cfa)
              DW_AT_call_all_tail_calls	(true)
              DW_AT_sibling	(0x0000010d)

0x000000b6:   DW_TAG_formal_parameter
                DW_AT_name	("argc")
                DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                DW_AT_decl_line	(2)
                DW_AT_decl_column	(0x0e)
                DW_AT_type	(0x00000058 "int")
                DW_AT_location	(DW_OP_fbreg -36)

0x000000c3:   DW_TAG_formal_parameter
                DW_AT_name	("argv")
                DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                DW_AT_decl_line	(2)
                DW_AT_decl_column	(0x1a)
                DW_AT_type	(0x0000010d "char**")
                DW_AT_location	(DW_OP_fbreg -48)

0x000000d0:   DW_TAG_lexical_block
                DW_AT_low_pc	(0x0000000000401146)
                DW_AT_high_pc	(0x000000000040115a)
                DW_AT_sibling	(0x000000f0)

0x000000e5:     DW_TAG_variable
                  DW_AT_name	("a")
                  DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                  DW_AT_decl_line	(4)
                  DW_AT_decl_column	(16)
                  DW_AT_type	(0x00000058 "int")
                  DW_AT_location	(DW_OP_reg3 RBX)

0x000000ef:     NULL

0x000000f0:   DW_TAG_lexical_block
                DW_AT_low_pc	(0x0000000000401164)
                DW_AT_high_pc	(0x0000000000401178)

0x00000101:     DW_TAG_variable
                  DW_AT_name	("a")
                  DW_AT_decl_file	("/home/wcohen/research/profiling/dyninst/pr1225/pr1225.c")
                  DW_AT_decl_line	(6)
                  DW_AT_decl_column	(16)
                  DW_AT_type	(0x00000058 "int")
                  DW_AT_location	(DW_OP_reg3 RBX)

0x0000010b:     NULL

0x0000010c:   NULL
```

**Comment by kupsch** [comment]
@wcohen thanks for the report.  This is fixed in PR #1226 on the branch kupsch/fix-variable-ranges. The range for variables is now properly scoped and your cse program now produces expected results.

**Comment by hainest** [comment]
@wcohen @kupsch Is it possible to add _some_ type of automated test for this? It wouldn't have to go in the test suite. We could stick it in https://github.com/dyninst/external-tests.

**Comment by wcohen** [comment]
Thinking that could further simplify the reproducer to have a single DW_OP_lexical_block and have a test program fetch the variable and check that it's range isn't [0,0xffffffffffffffff) as a test.  What is the difference between the dyninst external-tests and testsuite repositories?

**Comment by hainest** [comment]
>Thinking that could further simplify the reproducer to have a single DW_OP_lexical_block and have a test program fetch the variable and check that it's range isn't [0,0xffffffffffffffff) as a test.

That is certainly a necessary check. It would be great if we could independently find the address range of `main` and then check that the ranges of the variable are bounded by that. I wouldn't hold off on creating the more basic test, if we can't get that working, though.

>What is the difference between the dyninst external-tests and testsuite repositories?

No one knows how to add new mutatees to the test suite. We have tried several times to untangle the build machinery that controls that functionality, but we've never made it all the way through- we always get busy working on something more pressing. I created the external-tests repo as a stopgap so that we can at least have some way of doing more functionality testing. It's designed more for unit-like tests, as well.
